DBZ-7223 Add the MongoDB sink connector

rk3rn3r opened the pull request:
Todos:

- [x] Add tests
- [x] Optional: Add MongoDB source record format handling, filed [DBZ-7998](https://issues.redhat.com/browse/DBZ-7998)
- [x] Create Jira to add support for Debezium types (VariableScaleDecimal, MicroTime, NanoTime, ZonedTimestamp, ZonedTime, ...), filed [DBZ-7999](https://issues.redhat.com/browse/DBZ-7999)
- [x] Create Jira to add filter fields: include/exclude lists and tests, filed [DBZ-8000](https://issues.redhat.com/browse/DBZ-8000)
- [x] Create Jira to add SSL server API support, filed [DBZ-7995](https://issues.redhat.com/browse/DBZ-7995)
- [x] Create Jira for DLQ error handling support and tests (tolerateErrors), filed [DBZ-8001](https://issues.redhat.com/browse/DBZ-8001)

closes https://issues.redhat.com/browse/DBZ-7223

rk3rn3r commented:
@jpechane @jcechace @Naros FYI

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/SinkConnection.java:
```diff
@@ -0,0 +1,228 @@
+
+import io.debezium.config.Field;
+
+public final class SinkConnection {
```
Could this be refactored to not use just static methods?

rk3rn3r replied:
@jpechane Hmmm, I think a better class name would be `SinkMongoClientFactory` or something instead. This class/instance does not really reflect anymore its initial intention. wdyt?

jcechace replied:
There is one huge caveat in creating new layer of connection management for sink -- even considering the Source one needs some refactoring. You are losing the ability to use SSL connection. 

Java client for Mongo doesn't support client certificate configuration options (so things like `tlsCertificateKeyFile` etc...) and these need to be handled programatically...

rk3rn3r replied:
Agreed, and I still need to file a Jira to add SSL support as it might be important for many (cloud) scenarios.
@jcechace so what is your final conclusion? Keep it (maybe renamed) with static intializers or start a refactoring to a non-static/real `SinkConnection` class similar to the ones in the source connectors?  I need to finish this initial implementation PR this week because my PTO comes up then.

jcechace replied:
@rk3rn3r TBH it's up to @jpechane  in what form he is willing to merge. I personally think that SSL support is important and thus we should  consolidate the connection -- e.g. if possible reuse the one from source now and extend / refactor later. However since we are now in the 3.0 development cycle anyway I don't have a strong objection against doing this change in several PRs.

jpechane replied:
It is not ideal but I am willing to postpone it for full refactoring.

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/NoStackTraceDlqException.java:
```diff
@@ -0,0 +1,31 @@
+ * The length of the {@linkplain #getStackTrace() stack trace} is always 0, which means that this
+ * exception is suitable for reporting only situations with the stack trace being inconsequential.
+ */
+abstract class NoStackTraceDlqException extends DataException {
```
Why can't we keep the stack trace? Even if it seems to be inconsequential is most of the cases it can still be useful in others. What do we gain by dropping the stacktrace?

rk3rn3r replied:
This is meant for a custom error handling override to choose other options than the default behavior that comes for Kafka Connect sink connectors. I might remove this for now and file a Jira for adding it later. How is this going to be with Debezium Server for sink connectors? DLQs and sink error handling?

Naros replied:
That doesn't seem to answer the question though, does Kafka Connect mandate the use of `DataException` for these use cases or can it be any runtime exception ?

rk3rn3r replied:
@Naros Because you asked about that below. ;)

rk3rn3r replied:
I am going to remove this class for now. It is related to overriding the default Kafka Connect sink connector error handling with DLQ etc. I am going to file a Jira to add this later. It is not needed for the initial implementation. I hope to get it fully in but I think as my PTO comes up I can move this to a Jira and we can discuss and add it later. wdyt?

rk3rn3r replied:
I removed this class and related code for now (including `ErrorAnalyzer`). Those are meant to have a custom error handling overriding the default Kafka Connect DLQ/non-DLQ/error-tolerance error handling and giving more fine-constraint options to users. As those exception go as message into the DLQ the stack trace is about to be removed to not blow up message size. I will file a Jira for the overriding error handler behavior and keep the files locally until I re-add it again properly.

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/SinkRecordConverter.java:
```diff
@@ -0,0 +1,14 @@
+import org.apache.kafka.connect.data.Schema;
+import org.bson.BsonDocument;
+
+public interface SinkRecordConverter {
```
Please add JavaDoc.

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/SinkConverter.java:
```diff
@@ -0,0 +1,77 @@
+import io.debezium.connector.mongodb.converters.sink.LazyBsonDocument.Type;
+import io.debezium.table.ColumnNamingStrategy;
+
+public class SinkConverter {
```
Can we use a different name? The difference between `SinkConverter` abs `SinkRecordConverter` is not clear from naming. Esp if this class also operates on `SinkRecord`.

rk3rn3r replied:
`SinkConverter` became `SinkRecordConverter` as it is responsible to determine the correct `SinkValueConverter` based on the Kafka Connect SinkRecord/record.
`SinkRecordConverter` becomes `SinkValueConverter` interface that defines how values get converted into the MongoDB BsonDocument representing the original value.
done.

jpechane replied:
OK

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/SinkConverter.java:
```diff
@@ -0,0 +1,77 @@
+
+import io.debezium.connector.mongodb.converters.sink.LazyBsonDocument.Type;
+import io.debezium.table.ColumnNamingStrategy;
+
```
Please add JavaDoc.

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/FieldType.java:
```diff
@@ -0,0 +1,21 @@
+
+import org.apache.kafka.connect.data.Schema;
+
+public abstract class FieldType {
```
Do we need this as a separate class?

rk3rn3r replied:
Merged it with `AbstractBsonType` now.

jpechane replied:
OK

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/types/bson/BytesType.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Override
+    public BsonValue toBson(final Object data) {
```
We can test Java pattern matching on this connector In Debezium 3.x when we swtich to Java 17.

rk3rn3r replied:
Agreed! +1 definitely!

jpechane commented on debezium-connector-mongodb/src/main/resources/io/debezium/connector/mongodb/sink/build.version:
```diff
@@ -0,0 +1 @@
+version=${project.version}
```
Do we need a separate build version for sink connector when the source and sink ar ein the same package?

rk3rn3r replied:
No. Agreed.

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/MongoDbSinkConnector.java:
```diff
@@ -0,0 +1,77 @@
+
+public class MongoDbSinkConnector extends SinkConnector {
+
+    private static final List<String> REQUIRED_SINK_ACTIONS = Arrays.asList("insert", "update", "remove");
```
What's purpose of this attribute?

rk3rn3r replied:
Not used anymore after my previous refactoring. Removed.

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/Module.java:
```diff
@@ -0,0 +1,37 @@
+     * @return symbolic name of the connector plugin
+     */
+    public static String name() {
+        return "MongoDbSink";
```
I think we use lowercase here for all other connectors

rk3rn3r commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/event_handler/EventHandler.java:
```diff
@@ -0,0 +1,67 @@
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.mongodb.sink.event_handler;
```
But it's valid and not in the styleguide. So I think I can leave it? ;)

jcechace replied:
It's definitively unusual and I too think it looks weird. I'm not objecting per se. However, I'd suggest looking at whether we already have a "multi-word" package names already in the code base and use the same style in order to be consistent

jpechane replied:
Well, thanks for the reminder. We are going to add this into the styleguide as something that should not be used :-)

rk3rn3r replied:
Isn't this kind of bias based comment? I would personally find `record_emitter` much more readable than `recordemitter` especially when you are new to the code, not a native speaker or not aware of the specific context.
The official Java docs from Oracle even recommend underscores in package names when those originate from real domain names that have dashes in the name:
https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html

I think only Google/Google's Java styleguide prefers concatenation.

Naros replied:
Perhaps it may make more sense to find a better package name to avoid the need to even discuss this point.  We have packages like `converters` and `types`, which are super generic; why not just use `handlers` or some other meaning _word_ and avoid the use of two words altogether?

jcechace replied:
You are right, it is. So we need to reach a quorum -- or have Jiri overrule as democracy gets you only so far :).

jpechane replied:
@rk3rn3r Thanks

jcechace commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/SinkConnection.java:
```diff
@@ -0,0 +1,228 @@
+
+import io.debezium.config.Field;
+
+public final class SinkConnection {
```
A future jira should be logged to have a single MongoDB connection/client wrapper for both Sink and Source. 

It's ok for now as the connection package in source still needs revisiting and refactoring

rk3rn3r replied:
filed https://issues.redhat.com/browse/DBZ-7995

jcechace replied:
Just a note -- the cluster listener thingy has nothing to do with Atlas. I think I explained why it is used in the other comment.

jpechane replied:
@rk3rn3r Could you please updated the Jira description?

jcechace reviewed (commented):
Few comments for now. One thing which definitively needs to be used more are JavaDocs.

Naros commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/LazyBsonDocument.java:
```diff
@@ -0,0 +1,201 @@
+                    }
+                    break;
+                default:
+                    throw new DataException(format("Unknown data type %s.", dataType));
```
Would prefer use of `DebeziumException` instead.

Naros commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/SinkConverter.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    private SinkRecordConverter getConverter(final Schema schema, final Object data) {
+        if (schema != null && data instanceof Struct) {
+            LOGGER.debug("Using schema-ed converter");
+            return SCHEMA_RECORD_CONVERTER;
+        }
+
+        // structured JSON without schema
+        if (data instanceof Map) {
+            LOGGER.debug("Using schemaless / map converter");
+            return MAP_RECORD_CONVERTER;
+        }
+
+        // JSON string
+        if (data instanceof String) {
+            LOGGER.debug("Using JSON string converter");
+            return JSON_STRING_RECORD_CONVERTER;
+        }
+
+        // BSON bytes
+        if (data instanceof byte[]) {
+            LOGGER.debug("Using BSON converter");
+            return BYTE_ARRAY_RECORD_CONVERTER;
+        }
```
Why not make use of `if/elseif/else` semantics here?

jcechace replied:
Here I will object... Either use switch or leave it like this. I't definitively more readable IMHO. Once we switch to java 17 switch is likely to be better anyway.

rk3rn3r replied:
As there is immediate return  and the compile should handle it its own way anyway, I think I'll keep it like that until there's another call from @jpechane .

jpechane replied:
@rk3rn3r Keep it as is. We are on Java 17 so might bet better to handle it with pattern matching as thi final solution.
Could you please create a Jira where you record all instances in the connector that would benefit from Java 17 code refactoring?

rk3rn3r replied:
done. filed https://issues.redhat.com/browse/DBZ-8006 for it

Naros commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/SinkDocument.java:
```diff
@@ -0,0 +1,33 @@
+        this.valueDoc = valueDoc;
+    }
+
+    public Optional<BsonDocument> getKeyDoc() {
+        return Optional.ofNullable(keyDoc);
+    }
+
+    public Optional<BsonDocument> getValueDoc() {
+        return Optional.ofNullable(valueDoc);
+    }
```
I am not opposed to the use of `Optional`, but I don't care for the repeated creation of the `Optional` temporary objects for the sole purpose of nullability checks.  

If we can move the creation of the `Optional` to the constructor, then that'd be great.  If that presents an issue, then I think I'd prefer we use `!= null` or `Objects.isNull` instead.

jcechace replied:
I'm actually in favour of using Optional. However I agree with @Naros -- just make the field Optional right away. I know IDE is going to complain, I know the language designers argued it should not be done... And it's being done anyway since they never actually managed to explain properly why and IMHO it makes more sense.

rk3rn3r replied:
@Naros  What's the difference in numbers of `Optional` instances? Isn't the current implementation better because the lifespan is the lifespan of the variable that calls the `get*Doc()` methods? This can be garbage collected early. When the Optional is bound to the instance of `SinkDocument` then the lifespan might be pretty much longer, no?
This is why I think the authors suggest `Optional` to be used for return-methods only and not for member-vars and that's why IDEs should complain.

jcechace replied:
Being concerned about 2 extra references in an application such as Debezium seems a shining example of premature optimisation. I wouldn't even use the fact that the on call object construction is probably worse than the additional memory consumption. It's imho just a matter of readability TBH.

rk3rn3r replied:
Then I hope we can leave it like this for now? @jpechane ?
It would make my feelings better and my IDE's as well. ;)

jpechane replied:
We do use `Optional` for attributes and from my PoV it is good defensive programming approach.

rk3rn3r replied:
The rest of the world AND the authors for `Optional` do not agree though. Please file a Jira or change it yourself, I can't do it because to me it just does not feel right at all for the above reasons.

jpechane replied:
Code changed, issue resolved

Naros commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/SinkConverter.java:
```diff
@@ -0,0 +1,77 @@
+    private static final SinkRecordConverter BYTE_ARRAY_RECORD_CONVERTER = new ByteArrayRecordConverter();
+
+    public SinkDocument convert(final SinkRecord record, ColumnNamingStrategy columnNamingStrategy) {
+        LOGGER.debug("record: {}", record);
```
This class makes use of debug logging but similar ones use trace. Can we pick one and be consistent?

jpechane replied:
Please use `io.debezium.util.Loggings` whenever the data could be printed to the log.

rk3rn3r replied:
@Naros  This is intended. While `debug` level gives you good debug logging info the `trace` level gives you very detailed additional info to figure out where in the code what was happening. For general error finding `debug` will be good enough level and I want to keep much more details on the `trace` level which is used only in the data type conversion to get much more details on every processing step. It helped me a lot on the data type mapping change 3 weeks ago.

@jpechane is this mandatory for normal log levels or for debug/trace level too?

jpechane replied:
@rk3rn3r For security reasons we don't log the content of messages in the level other than trace. This came multiple time as a security concern from different community members.

rk3rn3r replied:
Understood! done

Naros commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/types/bson/BytesType.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Override
+    public BsonValue toBson(final Object data) {
+        if (data instanceof ByteBuffer) {
+            return new BsonBinary(((ByteBuffer) data).array());
+        }
+        if (data instanceof byte[]) {
+            return new BsonBinary((byte[]) data);
+        }
```
To be consistent, lets use `if/elseif/else`.

jcechace replied:
Why though? if-elseif-else makes sense in case of branching statement with consecutive code to be executed. When it's essentially an expression switch based on types (which should be used here once we have java 17) this seems more readable.

jpechane replied:
OK, let's make a decision. It is not sure we'll get this into 2.7. Even if we do we already have ibmi connector requiring Java 17. So I propose to change all the code to use switch/pattern matching. It will be kind of prototype for us for to build a code style around it.

jpechane replied:
Hold on, this is not possible with 2.7 as it is the same codebase with MongoDB source connector :-(

Naros commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/types/bson/DecimalType.java:
```diff
@@ -0,0 +1,51 @@
+    @Override
+    public BsonValue toBson(final Object data) {
+        if (data instanceof BigDecimal) {
+            if (format.equals(Format.DECIMAL128)) {
+                return new BsonDecimal128(new Decimal128((BigDecimal) data));
+            }
+            if (format.equals(Format.LEGACYDOUBLE)) {
+                return new BsonDouble(((BigDecimal) data).doubleValue());
+            }
```
Another if/elseif or perhaps switch use case.

jcechace replied:
Same response as in other cases :)... +1 for switch

rk3rn3r replied:
filed https://issues.redhat.com/browse/DBZ-8006 for the Java 17 enhancements/benefits

jpechane replied:
Leave it as is now and add it to the refactorig Jira.

Naros commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/MongoDbSinkConnectorConfig.java:
```diff
@@ -0,0 +1,234 @@
+            .withDescription("The name of the MongoDB database to which the connector writes to.")
+            .required();
+
+    public static final Field TABLE_NAME_FORMAT_FIELD = Field.create(TABLE_NAME_FORMAT)
+            .withDisplayName("A format string for the table")
+            .withType(ConfigDef.Type.STRING)
+            .withGroup(Field.createGroupEntry(Field.Group.CONNECTOR, 3))
+            .withWidth(ConfigDef.Width.MEDIUM)
+            .withImportance(ConfigDef.Importance.MEDIUM)
+            .withDefault("${topic}")
+            .withDescription("A format string for the table, which may contain '${topic}' as a placeholder for the original topic name.");
+
+    public static final Field TABLE_NAMING_STRATEGY_FIELD = Field.create(TABLE_NAMING_STRATEGY)
+            .withDisplayName("Name of the strategy class that implements the TablingNamingStrategy interface")
+            .withType(ConfigDef.Type.CLASS)
+            .withGroup(Field.createGroupEntry(Field.Group.CONNECTOR_ADVANCED, 2))
+            .withWidth(ConfigDef.Width.LONG)
+            .withImportance(ConfigDef.Importance.LOW)
+            .withDefault(DefaultTableNamingStrategy.class.getName())
+            .withDescription("Name of the strategy class that implements the TableNamingStrategy interface.");
+
+    public static final Field COLUMN_NAMING_STRATEGY_FIELD = Field.create(COLUMN_NAMING_STRATEGY)
+            .withDisplayName("Name of the strategy class that implements the ColumnNamingStrategy interface")
+            .withType(ConfigDef.Type.CLASS)
+            .withGroup(Field.createGroupEntry(Field.Group.CONNECTOR_ADVANCED, 2))
+            .withWidth(ConfigDef.Width.LONG)
+            .withImportance(ConfigDef.Importance.LOW)
+            .withDefault(DefaultColumnNamingStrategy.class.getName())
+            .withDescription("Name of the strategy class that implements the ColumnNamingStrategy interface.");
+
+    public static final Field BATCH_SIZE_FIELD = Field.create(BATCH_SIZE)
+            .withDisplayName("Specifies how many records to attempt to batch together into the destination table, when possible. " +
+                    "You can also configure the connector’s underlying consumer’s max.poll.records using consumer.override.max.poll.records in the connector configuration.")
+            .withType(ConfigDef.Type.INT)
+            .withGroup(Field.createGroupEntry(Field.Group.CONNECTOR_ADVANCED, 4))
+            .withWidth(ConfigDef.Width.SHORT)
+            .withImportance(ConfigDef.Importance.MEDIUM)
+            .withDefault(500)
+            .withDescription("Specifies how many records to attempt to batch together into the destination table, when possible. " +
+                    "You can also configure the connector’s underlying consumer’s max.poll.records using consumer.override.max.poll.records in the connector configuration.");
```
Maybe some of these fields that are shared with the JDBC sink could be moved to the parent class?

rk3rn3r replied:
We have plan for 2 options:
- move the JDBC sink to the main repo (and with this prepare a shared module for shared config)
- provide a shared module (or package inside debezium-core) with future shared sink config (`io.debezium.connector.SinkConnectorConfig` is the starting point introduced for this already)

I filed [DBZ-8002](https://issues.redhat.com/browse/DBZ-8002) for that.

jpechane replied:
Yes, let's do it together with code move.

Naros commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/event_handler/rdbms/RdbmsDeleteEvent.java:
```diff
@@ -0,0 +1,40 @@
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.mongodb.sink.event_handler.rdbms;
```
A nitpick, but I'd prefer we use the term `relational` rather than `rdbms`, just to remain consistent with other classes and packages that are associated with relational databases.  

This should also apply to class naming, not just the package name.

rk3rn3r replied:
It's not nitpick but a great syntactical comment!
done

Naros commented on debezium-core/src/main/java/io/debezium/table/TableNamingStrategy.java:
```diff
@@ -0,0 +1,24 @@
+     * @param record Kafka sink record, should not be {@code null}
+     * @return the resolved logical table name; if {@code null} the record should not be processed
+     */
+    String resolveTableName(SinkConnectorConfig config, org.apache.kafka.connect.sink.SinkRecord record);
```
```suggestion
    String resolveTableName(SinkConnectorConfig config, SinkRecord record);
```
And add the import to the import list.

rk3rn3r replied:
done

Naros reviewed (commented):
A few things I saw looking over this.  I'll try and give this a more thorough review this weekend.

jcechace commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/types/bson/BytesType.java:
```diff
@@ -0,0 +1,32 @@
+            return new BsonBinary((byte[]) data);
+        }
+
+        throw new DataException("Bytes field conversion failed due to unexpected object type " + data.getClass().getName());
```
Oh.. I assumed that DataException is expected by KC or something. IF that's not the case I definitively agree with @Naros 

The less KC API coupling we can have the better.

rk3rn3r replied:
Can I add a DebeziumDataException? That one is not related to runtime/environment related errors but purely data error related.

jpechane replied:
Please create a Jira for it - it seems it needs deeper analysis then.

rk3rn3r replied:
Filed [DBZ-8005](https://issues.redhat.com/browse/DBZ-8005)

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/AnalyzedBatchFailedWithBulkWriteException.java:
```diff
@@ -0,0 +1,133 @@
+import org.apache.kafka.connect.sink.SinkRecord;
+
+import com.mongodb.MongoBulkWriteException;
+import com.mongodb.assertions.Assertions;
```
> This class is not part of the public API and may be removed or changed at any time.

rk3rn3r replied:
class entirely removed

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/AnalyzedBatchFailedWithBulkWriteException.java:
```diff
@@ -0,0 +1,133 @@
+import com.mongodb.bulk.WriteConcernError;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.model.BulkWriteOptions;
+import com.mongodb.lang.Nullable;
```
Should we add our own annotation to `io.debezium.annotation`?

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/MongoDbSinkConnectorTask.java:
```diff
@@ -0,0 +1,140 @@
+import io.debezium.dlq.ErrorReporter;
+
+public class MongoDbSinkConnectorTask extends SinkTask {
+    static final Logger LOGGER = LoggerFactory.getLogger(MongoDbSinkConnectorTask.class);
```
```suggestion
    private static final Logger LOGGER = LoggerFactory.getLogger(MongoDbSinkConnectorTask.class);
```

rk3rn3r replied:
No, this is actually needed as the sub-task `io.debezium.connector.mongodb.sink.StartedMongoDbSinkTask` aggregates its logs under this `MongoDbSinkConnectorTask` logger.

jpechane replied:
@rk3rn3r In this case I'd prefere `protected`

rk3rn3r replied:
I think that won't work as there is not inheritance. That's why it's `package-private`, which is imo even stricter than protected, no?

jpechane replied:
Well, your comment says `sub-class` that's why I expected the inheritance :-)

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/AnalyzedBatchFailedWithBulkWriteException.java:
```diff
@@ -0,0 +1,133 @@
+        }
+    }
+
+    public interface Logger {
```
Could we have a different/more specific name? `Logger` is too associated with log4j.

rk3rn3r replied:
Removed

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/SinkConverter.java:
```diff
@@ -0,0 +1,77 @@
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.mongodb.converters.sink;
```
```suggestion
package io.debezium.connector.mongodb.sink.converters;
```

I don't think we should mix different converter types into the same namespace.

rk3rn3r replied:
done

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/types/bson/DecimalType.java:
```diff
@@ -0,0 +1,51 @@
+
+    public enum Format {
+        DECIMAL128, // for MongoDB v3.4+
+        LEGACYDOUBLE // results in double approximation
```
I agree. Is there a way how Debezium can receive kegacy double on supported MongoDB versions?

jcechace replied:
> I agree. Is there a way how Debezium can receive kegacy double on supported MongoDB versions?

In this context the enum represents the target type used for BigDecimal. SO I guess the short answer is "No"...  

Prior to 3.4 mongo didnt have a dedicated Decimal type, so you would only use doubles there. And doubles would be doubles for debezium.

rk3rn3r replied:
@jcechace which one to be removed precisely? `LEGACYDOUBLE`, `DECIMAL128` or both entirely?

jcechace replied:
@jcechace  well LEGACYDOUBLE is to be removed as it's not really applicable (we don't support such ancient MongoDB version). Consequently there is no point in having just a single option, right? So that means DECIMAL128 will become the only (implicit) variant.

jpechane replied:
Yes, IIUC the enum can be removed completely.

rk3rn3r replied:
done in an extra commit

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/WriteModelStrategy.java:
```diff
@@ -0,0 +1,17 @@
+
+import io.debezium.connector.mongodb.converters.sink.SinkDocument;
+
+public interface WriteModelStrategy {
```
Please add JavaDoc

rk3rn3r replied:
This is meant for a single default implementation and the interface might eventually change. Leaving it for the future strategies to add javadoc. Is this okay?

Filed https://issues.redhat.com/browse/DBZ-7997 for it.

jcechace replied:
I really think that we shouldn't have undocumented interfaces at all.  It simply shouldn't be a thing in our code base.

jpechane replied:
I think short description of the intent does not hurt. You don't need to document the structure of the interface.

rk3rn3r replied:
done in an extra commit

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/DefaultWriteModelStrategy.java:
```diff
@@ -0,0 +1,26 @@
+
+import io.debezium.connector.mongodb.converters.sink.SinkDocument;
+
+public class DefaultWriteModelStrategy implements WriteModelStrategy {
```
Do we need a separate class?

rk3rn3r replied:
Yes, there are other scenarios to be implemented: Insert/Append-only, Update (the doc content / changed fields instead of replacing the whole document), Delete, etc

Filed https://issues.redhat.com/browse/DBZ-7997 for that.

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/IdStrategy.java:
```diff
@@ -0,0 +1,16 @@
+
+import io.debezium.connector.mongodb.converters.sink.SinkDocument;
+
+public interface IdStrategy {
```
Please add JavaDoc. Is this something tht is intended to be configurable in future using custom class?

rk3rn3r replied:
done

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/WriteModelStrategyHelper.java:
```diff
@@ -0,0 +1,52 @@
+
+import io.debezium.connector.mongodb.converters.sink.SinkDocument;
+
+public final class WriteModelStrategyHelper {
```
Is this class even used?

rk3rn3r replied:
I refactored this part. It's not in use anymore.

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/StartedMongoDbSinkTask.java:
```diff
@@ -0,0 +1,154 @@
+            }
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
```
Use `DebeziumException` where possible instead of `RuntimeException`

rk3rn3r replied:
done

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/StartedMongoDbSinkTask.java:
```diff
@@ -0,0 +1,154 @@
+
+import io.debezium.dlq.ErrorReporter;
+
+final class StartedMongoDbSinkTask implements AutoCloseable {
```
Is it possible to use `io.debezium.pipeline.sink.spi.ChangeEventSink`?
I can envision then `BaseSinkTask` that will manage the sink connector/task lifecycle.

jpechane replied:
Also we should intorduce `BaseSinkConnector` even if it is empty at the start.

rk3rn3r replied:
Possibly yes, I will log a Jira for it. Looking at my upcoming PTO I want to merge it like that and refactor it with an extra task later. is this fine for you @jpechane ?

rk3rn3r replied:
filed https://issues.redhat.com/browse/DBZ-7996

rk3rn3r commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java:
```diff
@@ -1440,7 +1440,7 @@ private static int resolveSnapshotMaxThreads(Configuration config) {
     }
 
-    private static ConnectionString resolveConnectionString(Configuration config) {
+    public static ConnectionString resolveConnectionString(Configuration config) {
```
I think the behavior is quite general and although from the outsides-outside it is internal, but resolving the conn string from the config is quite a reusable logic? Eventually this could go to a kind of helper class maybe? Especially as sink and source connectors are in the same Maven module/jar. wdyt?

jcechace replied:
So move it into that Unit method, please. Like this it's wrong.

rk3rn3r replied:
see io.debezium.connector.mongodb.shared.SharedMongoDbConnectorConfig

rk3rn3r replied:
done

rk3rn3r commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/AnalyzedBatchFailedWithBulkWriteException.java:
```diff
@@ -0,0 +1,133 @@
+ * the DLQ a {@linkplain MongoCollection#bulkWrite(List, BulkWriteOptions) batch} failed with {@link
+ * MongoBulkWriteException}.
+ */
+public final class AnalyzedBatchFailedWithBulkWriteException {
```
Good point. Are you fine with  `ErrorAnalyzer`?

jcechace replied:
Sure

rk3rn3r replied:
Done (with the next push)

rk3rn3r commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/SinkConnection.java:
```diff
@@ -0,0 +1,228 @@
+            ConnectionString connectionString = new ConnectionString((String) configValue.value());
+            MongoClientSettings mongoClientSettings = MongoClientSettings.builder()
+                    .applyConnectionString(connectionString)
+                    .applyToClusterSettings(
+                            b -> b.addClusterListener(
+                                    new ClusterListener() {
+                                        @Override
+                                        public void clusterOpening(final ClusterOpeningEvent event) {
+                                        }
+
+                                        @Override
+                                        public void clusterClosed(final ClusterClosedEvent event) {
+                                        }
+
+                                        @Override
+                                        public void clusterDescriptionChanged(
+                                                                              final ClusterDescriptionChangedEvent event) {
+                                            ReadPreference readPreference = connectionString.getReadPreference() != null
+                                                    ? connectionString.getReadPreference()
+                                                    : ReadPreference.primaryPreferred();
+                                            if (!connected.get()
+                                                    && event.getNewDescription().hasReadableServer(readPreference)) {
+                                                connected.set(true);
+                                                latch.countDown();
+                                            }
+                                        }
+                                    }))
+                    .build();
```
I saw this when I studied other Mongo DB connectors. I think this is meant for MongoDB Atlas clusters and handling that nodes get removed or added? I don't know and not like M much, so I am not 100% sure.

jcechace replied:
I can imagine the use in cases, where you need to e.g. restart connection when cluster topology would change. I intended to use something very similar for killing change stream in case of a promotion event.  

However here it seems to be a one time use -- meaning any attempt to communicate with the database would do the same.

Regardless, the information that other use it is valuable and  I will try to have a look at what e.g. the official Mongodb connector does and why they would use that in case such as this one. 

BTW: I haven't personally verified (it is on my toto list), however based on what Bobby told me, the event is not fired in case of added replica to an existing shard.

jcechace replied:
Ok, so this seems to be done this way primarily for two reason 

1) At this point we don't know what permissions we are missing (so forcing the connection with e.g. listing databases might not work)
2) It somewhat has the benefit of not really interacting with the database any more than we need to. 

We should probably utilise similar approach even in source.

rk3rn3r replied:
Agreed. I will file a Jira for it!

rk3rn3r replied:
added to https://issues.redhat.com/browse/DBZ-7995

rk3rn3r commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/SinkConverter.java:
```diff
@@ -0,0 +1,77 @@
+    private static final SinkRecordConverter JSON_STRING_RECORD_CONVERTER = new JsonStringRecordConverter();
+    private static final SinkRecordConverter BYTE_ARRAY_RECORD_CONVERTER = new ByteArrayRecordConverter();
+
+    public SinkDocument convert(final SinkRecord record, ColumnNamingStrategy columnNamingStrategy) {
```
This is outdated. Sorry.

jcechace commented:
@rk3rn3r Since this is a larger PR, could you please not use the "Mark as resolved" functionality at least until after things are really resolved in the code -- preferably as a standalone commit (we can always squash later). Though I'd actually not use it at all. It makes the changes much easier to follow.

Thank you. 

@jpechane @Naros  What's your take on this? I personally find it difficult to track when this is used (in case of larger PR, unless the discussion is already from some old iteration). Unfortunately it seems that GH wont even notify in such cases.

jpechane commented:
My take is that `Resolve` should be done by the author of the comment as it is up to him to understand the exaplanation.

Naros commented:
I've personally used it some in the past, but where I use it most often is in documentation PRs, but that workflow tends to be very different as Bob more times than not looks at the PR change in total to guarantee that the wording flows rather than trying to make that determination from commit to commit.

But what I find most important, particularly if the PR has been reviewed already, is to avoid a rebase unless necessary.  Sending new commits to the PR just makes future PR passes easier to follow for the reviewer.

rk3rn3r commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/types/bson/AbstractBsonType.java:
```diff
@@ -0,0 +1,48 @@
+
+import io.debezium.connector.mongodb.converters.sink.FieldType;
+
+public abstract class AbstractBsonType extends FieldType {
```
good point! done.

rk3rn3r commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/SinkConnection.java:
```diff
@@ -0,0 +1,228 @@
+        return Optional.empty();
+    }
+
+    public static void checkUserHasActions(
```
Agreed. Commented out for now and I will file a Jira for this.

rk3rn3r replied:
should be part of https://issues.redhat.com/browse/DBZ-7995

rk3rn3r commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/StartedMongoDbSinkTask.java:
```diff
@@ -0,0 +1,154 @@
+    @SuppressWarnings("try")
+    @Override
+    public void close() {
+        try (MongoClient autoCloseable = mongoClient) {
```
I will add comments in the code for this. And leave it like that.

jpechane replied:
You can keep it as is but still log the exception on the `DEBUG` level in case `close` throws one.

jpechane replied:
Also an exception in this phase should not get propagated.

rk3rn3r commented:
I addressed most of the comments with my recent changes to the branch. I hope we can get the PR through the door this week as my PTO is coming up and I like to finish this before. @jpechane

rk3rn3r commented:
> My take is that Resolve should be done by the author of the comment as it is up to him to understand the exaplanation.

I waited for this in many PRs prior this one and that never happened. Though I agree with this 1000% !

rk3rn3r commented:
> Unfortunately it seems that GH wont even notify in such cases.

True. I usually comment first and then resolve the discussion because of that.

rk3rn3r commented:
@jpechane @Naros @jcechace First of all thank you for the review. I addressed most of the concerns either by changing the code as requested or by filing a Jira task.
Please have a look  at the discussions that I left open and maybe review if my changes for the resolved discussions are acceptable.
I am adding a basic test case by tomorrow and I hope we can merge the PR then by Thu before my PTO starts next week. :pray:

jpechane commented:
@rk3rn3r Thanks, I belive we are almost done. Could you please address the comments I did recently? Thanks

jpechane commented:
@rk3rn3r Could you please run the code through the formatter too? Thanks

Naros commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/MongoProcessedSinkRecordData.java:
```diff
@@ -0,0 +1,91 @@
+        }
+        catch (Exception e) {
+            exception = e;
+            // if (config.logErrors()) {
+            LOGGER.error("Unable to process record {}", sinkRecord, e);
+            // }
+            // if (!config.tolerateErrors()) {
+            // throw e;
+            // }
```
Can we remove the superfluous comments here?

rk3rn3r replied:
Maybe yes, maybe no.
The code is needed soon when adding the error tolerance overrides.
I'll have a look how to handle this.

rk3rn3r replied:
done

Naros commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/SinkConnection.java:
```diff
@@ -0,0 +1,213 @@
+        return Optional.empty();
+    }
+
+//    public static void checkUserHasActions(
```
Is this still needed?

rk3rn3r replied:
later yes. same as above.

rk3rn3r replied:
done

Naros commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/sink/MongoDbSinkConnectorConfig.java:
```diff
@@ -0,0 +1,232 @@
+            .withDescription("Specifies how many records to attempt to batch together into the destination table, when possible. " +
+                    "You can also configure the connector’s underlying consumer’s max.poll.records using consumer.override.max.poll.records in the connector configuration.");
+
+    // public static final Field FIELD_INCLUDE_LIST_FIELD = Field.create(FIELD_INCLUDE_LIST)
```
Lets remove these commented bits if they're not being used.

rk3rn3r replied:
same.

rk3rn3r replied:
done

Naros reviewed (commented):
@rk3rn3r just a few inline bits.

rk3rn3r commented:
@jpechane I think this PR is basically ready to be merged. All tests passed.
I will add one commit to address the commented code, like we discussed this morning.

There is one test error for the Debezium Connect REST Extension failure which should be unrelated.
It seems to be caused by an old Java version?! 
`java.lang.UnsupportedClassVersionError: io/debezium/kcrestextension/DebeziumConnectRestExtension has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 55.0`
Any idea?

rk3rn3r commented:
@jpechane could you please not squash those commits? It would keep the different changes cleaner isolated and the code should work in all states / commits. It is also easier to bring back the functionalities that I prepared but that we moved to after my PTO / later.

jcechace commented:
> @jpechane could you please not squash those commits? It would keep the different changes cleaner isolated and the code should work in all states / commits. It is also easier to bring back the functionalities that I prepared but that we moved to after my PTO / later.

In that case, please do reword the commits and use meaningful commit messages. Otherwise I think @jpechane  should squash it.

jcechace commented on debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/junit/MongoDbDatabaseProvider.java:
```diff
@@ -45,15 +53,19 @@ public static MongoDbReplicaSet dockerReplicaSet() {
      */
     public static MongoDbReplicaSet dockerAuthReplicaSet() {
-        return dockerReplicaSetBuilder().authEnabled(true).build();
+        var replicaSet = dockerReplicaSetBuilder().authEnabled(true);
+        if (NETWORK != null) {
+            replicaSet.network(NETWORK);
+        }
+        return replicaSet.build();
```
There is literary no reason to touch this.

rk3rn3r replied:
There is no way when using this **** static abstraction layer to customize network. It's okay to have this layer, but even jumping out here into the MongoDbDeployment and back into those methods is not good. This is just a way to not throw away that layer and rewrite everything with such a PR. Please file a Jira for cleaning up this abstraction layer.  It is working for now and we are only talking about test code.

jcechace commented on debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/junit/MongoDbDatabaseProvider.java:
```diff
@@ -36,7 +40,11 @@ private static MongoDbReplicaSet.Builder dockerReplicaSetBuilder() {
     public static MongoDbReplicaSet dockerReplicaSet() {
-        return dockerReplicaSetBuilder().build();
+        var replicaSet = dockerReplicaSetBuilder();
+        if (NETWORK != null) {
+            replicaSet.network(NETWORK);
+        }
+        return replicaSet.build();
```
Pass the network as an argument if needs be and create a non parameter overrride. We are not using a static filed like this.

rk3rn3r replied:
That's not working because of timing issues. The instance is created before we can pass the network if we don't sneak it in like this for now. 
This MongoDbDatabaseProvider and MongoDbDeployment etc needs bigger refactoring it seems. It is not possible to do it in an easy, clean way. Do you have a suggestion? Something which is achievable by tomorrow? That's the timeline.

jcechace replied:
1) You don't actually need to use this class at all. It came to be due to series of refactorings and a need not to overhaul the entire existing test suite. 

2) That Version provider only really determines the version 
3) What seems to be the problem with `MongoDbDeployment`? 
4) I get the if, I just don't agree with it being based on the static attribute. I'd personally  change the 2 builder methods to just use the common network in case a null is passed to the builder.

rk3rn3r replied:
ad 1) What is the reason behind this class? I assume based on a system property you can switch the build to use external provided MongoDB/DocumentDB instance. If we want to support such tests, and I think it makes a lot of sense, it also makes sense to use this class.

ad 2) no, based on `mongodb.platform` system property you can switch to use an external instance, right?

ad 3) sorry, I meant `MongoDbPlatform`. It's used from `MongoDbDatabaseProvider` but `MongoDbPlatform` then calls code from `MongoDbDatabaseProvider` and there's no way to inject custom config.

ad 4) How you want to do that when using that abstraction layer that provides the switch for external instances? I think that's a possible temporary solution. 

I filed [DBZ-8017](https://issues.redhat.com/browse/DBZ-8017) for the proper refactoring.

jcechace replied:
1) Centralise the "creation" of Mongo deployment FOR THE MOST COMMON USE CASES IN THE SUITE - which among others counts on usage WITH EMBEDDED ENGINE where you never have the problem with network in the first place. 

You absolutely don't have to use this class -- e.g. the auth tests are not using it, since obviously only those require the database to be set with authentication and it wouldn't make sense to rewrite everything.  

2) Yes. 

3) In case of external you don't even need the custom config. Your entire problem is the docker. 


Your entire problem boils down to the fact that you can't pass state down to the provider/supplier used in 


```
 public static MongoDbDeployment externalOrDockerReplicaSet() {
        var platform = MongoDbDatabaseVersionResolver.getPlatform();
        return platform.provider.get();
    }
``

So you can either create and override and refactor MongoDbPlatform. Or just don't use this method and have your test just work with `dockerReplicaSet()` (or add the ability to pass in the network). 

Look, I agree that this is something which in ideal world would be implemented differently -- however it is like this for historical reason where we just didn't have the resources for a complete test rewrite (something we will have to do in the future anyway). However we should make it worse and introducing a potentially unstable side effect does make it worse. 

E.g. where is the `NETWORK` attribute set back to null? How do you ensure that the tests which will run after / intertwined with yous will not suddenly stat using that network?

rk3rn3r replied:
see below. after you comment about the side-effects I incorporated the reset of NETWORK.

Naros commented:
> There is one test error for the Debezium Connect REST Extension failure which should be unrelated.

@rk3rn3r speaking of this lovely class, I'd suggest not rename this class.  At the very least the rename should ideally be in a separate commit all by itself given that it touches all connectors and seems to be buried in a commit that really doesn't fully describe the breadth of that change.  When introducing a connector, I can expect there may be dependency changes in things like `debezium-core` for example, but I would not have expected any impacts of this PR to be on any other connector repository other than Mongo.

jcechace commented on debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/junit/MongoDbDatabaseProvider.java:
```diff
@@ -63,7 +75,20 @@ public static MongoDbShardedCluster mongoDbShardedCluster() {
+
+    public static MongoDbShardedCluster mongoDbShardedCluster() {
+        var cluster = mongoDbShardedClusterBuilder();
+        if (NETWORK != null) {
+            cluster.network(NETWORK);
+        }
```
Why do you need this when there is also the method which takes a network?

rk3rn3r replied:
Because the common/main entrypoint is `io.debezium.connector.mongodb.junit.MongoDbDatabaseProvider#externalOrDockerReplicaSet(org.testcontainers.containers.Network)` which is there to support external `DocumentDB`.
I think this needs to be cleaned up?

jcechace replied:
Unless you intend to test with Document DB (and in that case I'd like to see the actual test result, please). Then I don't see how that would be relevant. 

In case of external database the network is logically not needed at all. We certainly can revisit this to make thing more compact again. However I don't see a reason for these changes as part of this PR when for your intended purpose you can just call the specific method and provide the network directly.

rk3rn3r replied:
I assume it makes a lot of sense to test the sink against different providers eventually.

jcechace commented on debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/junit/MongoDbDatabaseProvider.java:
```diff
@@ -78,6 +103,17 @@ public static MongoDbDeployment externalOrDockerReplicaSet() {
+     * @return MongoDb replica-set deployment
+     */
+    public static MongoDbDeployment externalOrDockerReplicaSet(Network network) {
+        NETWORK = network;
```
This just doesn't seem like something we should do.

rk3rn3r replied:
The reason for this is that all those methods are already static. That's just the sugar on top of the cake, no?

jcechace replied:
That switch is not magical -- is purely based on property and can be backtraced in IDE easily. 

Additionally there is a significant difference between static method and static attribute -- just one of those creates a state.

rk3rn3r replied:
Not really. A static method, which in this case has no option to pass custom config changes / a custom builder or parameters is a shortcoming. I think it would be better to move out some instance config/creation code from `MongoDbDatabaseProvider` into the `MongoDbReplicaSet` and `MongoDbShardedCluster` classes and refactor `MongoDbPlatform` to allow passing a custom builder that reconfigures the default config when provided. That way that static intermediate state is not bad.

On the other hand: It's not too bad to have this static state for the Network in `MongoDbDatabaseProvider`. This class lives inside the `test` context only and I assume there's only one network per test run.

One suggestion:
```
    public static MongoDbReplicaSet dockerReplicaSet() {
        var replicaSetBuilder = dockerReplicaSetBuilder();
        if (NETWORK != null) {
            replicaSetBuilder.network(NETWORK);
        }
        var replicaSet = replicaSetBuilder.build();
        NETWORK = null;
        return replicaSet;
    }
```
Would that solve the side effects you mentioned?

rk3rn3r replied:
@jcechace wrote above:


> E.g. where is the `NETWORK` attribute set back to null? How do you ensure that the tests which will run after / intertwined with yous will not suddenly stat using that network?

see here. I added reset.

jcechace replied:
So you have a test class which sets the NETWORK, then you have a different test class which wants  a fresh new network / the default and so it calls `dockerReplicaSet()`. What happens? It gets the old one. Or am I reading it wrong?

BTW: this doesn't really solve anything, it just digs the pit of temporary workarounds even deeper.

rk3rn3r commented:
> At the very least the rename should ideally be in a separate commit

@Naros  it is a separate commit.

jpechane commented:
To move things forward

* Please move tests into a separate package `io.debezium.connector.mongodb.sink`
* Don't change existing MongoDB tests or infrastructure
* If any change is needed feel free to make a copy of existing classes, put it into `io.debezium.connector.mongodb.sink` and modify them as you need - to keep the changes isolated
* We'll merge the PR afterwards
* When I get back from PTO we'll sit together and unify and refactor the code as needed

jpechane commented on debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/SinkConnectorIT.java:
```diff
@@ -0,0 +1,122 @@
+                .until(() -> listCollections(DATABASE_NAME).size() >= 6);
+        List<String> collections = listCollections(DATABASE_NAME);
+        LOGGER.debug("List collections: {}", Arrays.toString(collections.toArray()));
+        Assertions.assertThat(listCollections(DATABASE_NAME).size()).isEqualTo(6);
```
@rk3rn3r I think we can use `assertj` only. There can be for example
```
Assertions.assertThat(listCollections(DATABASE_NAME)).hasSize(6);
        Assertions.assertThat(collections).containsExactlyInAnyOrder("dbserver1_inventory_addresses", "dbserver1_inventory_orders", "dbserver1_inventory_customers",
                "dbserver1_inventory_products_on_hand", "dbserver1_inventory_geom", "dbserver1_inventory_products");
```

rk3rn3r replied:
Why is that

> we can use `assertj` only

?

jpechane replied:
Wy should we use both hamcrest and assertj in a single class if they offer the same functionality?

rk3rn3r replied:
done

rk3rn3r commented:
>     * Please move tests into a separate package `io.debezium.connector.mongodb.sink`

I don't think that's an appropriate solution. The `MongoDbSinkConnector` class should end up in the package `io.debezium.connector.mongodb` (root level), like we discussed in one of our calls to align it to the other connector that we were talking about. Then test namespace should be `io.debezium.connector.mongodb` too, right? 
Maybe I should move `MongoDbSinkConnector` from `io.debezium.connector.mongodb.sink` to ``io.debezium.connector.mongodb` already now with this PR to avoid the `connector.class` breaking config change later?`

> 
>     * Don't change existing MongoDB tests or infrastructure

What exactly do you mean? What I had to do is what I had to do. Please suggest more concrete changes else I don't know what do you expect. Please mind the deadline for all changes is today!

> 
>     * If any change is needed feel free to make a copy of existing classes, put it into `io.debezium.connector.mongodb.sink` and modify them as you need - to keep the changes isolated

see above. Furthermore I don't see value in copying too much now that can and should easily be shared with the source connector.

rk3rn3r commented:
@jpechane I finished the discussed changes. I also added the rest of NETWORK that I suggested above. Feel free to merge the PR like this or do the copying of classes.
From my side that copying is not needed and I think also @jcechace is not entirely happy with the way it is now (I am neither but it is an okay-ish solution that requires bigger refactoring later) and we can do the refactoring of those classes so it is possible to provide a custom builder properly without slipping it in that dirty way - fully agreed.
